Realm/RLMObjectBase.mm

```diff
@@ -256,12 +256,12 @@ - (BOOL)isEqualToObject:(RLMObject *)object {
 }
 
 - (BOOL)isEqual:(id)object {
-    if (_objectSchema.primaryKeyProperty) {
-        return [self isEqualToObject:object];
-    }
-    else {
-        return [super isEqual:object];
+    if (RLMObjectBase *other = RLMDynamicCast<RLMObjectBase>(object)) {
+        if (_objectSchema.primaryKeyProperty) {
+            return [self isEqualToObject:other];
+        }
     }
+    return [super isEqual:object];
```

Comment:
not necessarily, since we only check `isEqualToObject:` when there's a primary key property